File: backend-spring/src/main/java/com/nearkart/entity/Expense.java
package com.nearkart.entity;

import jakarta.persistence.*;
import lombok.*;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "expenses")
@Getter @Setter
@NoArgsConstructor @AllArgsConstructor
@Builder
public class Expense {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @Column(name = "company_id")
    private UUID companyId;

    @Column(name = "expense_date")
    private LocalDate expenseDate;

    private String category;

    @Column(columnDefinition = "TEXT")
    private String narration;

    @Column(nullable = false, precision = 14, scale = 2)
    private BigDecimal amount;

    @Column(name = "payment_mode")
    private String paymentMode;

    @Column(name = "bank_account_id")
    private UUID bankAccountId;

    @Column(name = "cheque_no")
    private String chequeNo;

    @Column(name = "party_name")
    private String partyName;

    @Column(name = "farmer_bill_ref")
    private String farmerBillRef;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        OffsetDateTime now = OffsetDateTime.now();
        createdAt = now; updatedAt = now;
        if (expenseDate == null) expenseDate = LocalDate.now();
    }

    @PreUpdate
    protected void onUpdate() { updatedAt = OffsetDateTime.now(); }
}

File: backend/app/schemas/expense.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class ExpenseCreate(BaseModel):
    company_id: UUID | None = None
    expense_date: date
    category: str | None = None
    narration: str | None = None
    amount: Decimal
    payment_mode: str = "cash"
    bank_account_id: UUID | None = None
    cheque_no: str | None = None
    party_name: str | None = None
    farmer_bill_ref: str | None = None


class ExpenseUpdate(BaseModel):
    company_id: UUID | None = None
    expense_date: date | None = None
    category: str | None = None
    narration: str | None = None
    amount: Decimal | None = None
    payment_mode: str | None = None
    bank_account_id: UUID | None = None
    cheque_no: str | None = None
    party_name: str | None = None
    farmer_bill_ref: str | None = None


class ExpenseResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    company_id: UUID | None
    expense_date: date
    category: str | None
    narration: str | None
    amount: Decimal
    payment_mode: str
    bank_account_id: UUID | None
    cheque_no: str | None
    party_name: str | None
    farmer_bill_ref: str | None
    created_at: datetime
    updated_at: datetime
